src/app/signup/page.tsx

```diff
@@ -1,22 +1,83 @@
 "use client";
 
-import Image from "next/image";
 import { GradButton } from "@/components/ui/grad-button";
 import SignUpImage from "@/components/ui/signup-image";
+import { signupUser } from "@/lib/User/actions";
+import {
+  EMAIL_ALREADY_EXISTS_MSG,
+  SIGNUP_MAX_EMAIL_LENGTH,
+  SIGNUP_MAX_NAME_LENGTH,
+  SIGNUP_MAX_PASSWORD_LENGTH,
+  SIGNUP_MIN_NAME_LENGTH,
+  SIGNUP_MIN_PASSWORD_LENGTH,
+  signUpSchema,
+} from "@/lib/User/validations";
 import { zodResolver } from "@hookform/resolvers/zod";
+import Image from "next/image";
+import { useMemo, useRef, useState } from "react";
+import { useFormState } from "react-dom";
 import { useForm } from "react-hook-form";
-import { signupUser } from "@/lib/User/actions";
-import { signUpSchema } from "@/lib/User/validations";
 
 const SignUp = () => {
+  const formRef = useRef<HTMLFormElement>(null);
+  const [state, formAction] = useFormState(signupUser, undefined);
+  const previousEmailValue = useMemo(() => {
```

Comment:
Can we consolidate these into a single function? Also, you likely want to do useCallback here instead of useMemo

Reply:
Sure. It should be pretty quick. But why useCallback though? We're only memoizing a value, aren't we?

Reply:
Memorizing that value may not result in the intended effect. I think useCallback here would make more sense. Typically useMemo is used for functions that take an input and you memorize off that input key